### DISCUSSION TITLE
PLEASE HELP: Last mile Block in implementing a GUI for rhvoice

### DISCUSSION THREAD

**Discussion by rajibando** [RHVoice/RHVoice#907] [Q&A]
Dear Team RHVoice,

I installed RHVoice list of packages as directed by Apt/Synaptic Package Manager.
My laptop's Debian was installed from `debian-live-12.6.0-amd64-lxqt.iso`.

The list of packages are as follows:

> Commit Log for Sat Nov 30 21:14:40 2024
> Installed the following packages:
librhvoice-audio2 (1.8.0+dfsg-3)
librhvoice-core7 (1.8.0+dfsg-3)
rhvoice (1.8.0+dfsg-3)
rhvoice-english (1.8.0+dfsg-3)

The complete list of dependencies (this I gathered with a script) are as follows:
> Package: rhvoice
Download-Size: 170 kB
Installed-Size: 640 kB
Dependencies:
  Dependency: libc6
Download-Size: 2,757 kB
Installed-Size: 13.3 MB
- - - - - -
  Dependency: libgcc-s1
Download-Size: 49.9 kB
Installed-Size: 143 kB
- - - - - -
  Dependency: libglib2.0-0
Download-Size: 1,400 kB
Installed-Size: 4,231 kB
- - - - - -
  Dependency: libglibmm-2.4-1v5
Download-Size: 607 kB
Installed-Size: 2,867 kB
- - - - - -
  Dependency: librhvoice-audio2
Download-Size: 39.7 kB
Installed-Size: 220 kB
- - - - - -
  Dependency: librhvoice-core7
Download-Size: 511 kB
Installed-Size: 2,399 kB
- - - - - -
  Dependency: libsigc++-2.0-0v5
Download-Size: 25.7 kB
Installed-Size: 69.6 kB
- - - - - -
  Dependency: libstdc++6
Download-Size: 614 kB
Installed-Size: 2,750 kB
- - - - - -
` = = = = ` = = = = ` = = = = ` = = = = =
Package: librhvoice-audio2
Download-Size: 39.7 kB
Installed-Size: 220 kB
Dependencies:
  Dependency: libao4
Download-Size: 37.1 kB
Installed-Size: 158 kB
- - - - - -
  Dependency: libc6
Download-Size: 2,757 kB
Installed-Size: 13.3 MB
- - - - - -
  Dependency: libgcc-s1
Download-Size: 49.9 kB
Installed-Size: 143 kB
- - - - - -
  Dependency: libportaudio2
Download-Size: 66.7 kB
Installed-Size: 219 kB
- - - - - -
  Dependency: libpulse0
Download-Size: 274 kB
Installed-Size: 992 kB
- - - - - -
  Dependency: librhvoice-core7
Download-Size: 511 kB
Installed-Size: 2,399 kB
- - - - - -
  Dependency: libstdc++6
Download-Size: 614 kB
Installed-Size: 2,750 kB
- - - - - -
` = = = = ` = = = = ` = = = = ` = = = = =
Package: librhvoice-core7
Download-Size: 511 kB
Installed-Size: 2,399 kB
Dependencies:
  Dependency: libc6
Download-Size: 2,757 kB
Installed-Size: 13.3 MB
- - - - - -
  Dependency: libgcc-s1
Download-Size: 49.9 kB
Installed-Size: 143 kB
- - - - - -
  Dependency: libsonic0
Download-Size: 11.0 kB
Installed-Size: 34.8 kB
- - - - - -
  Dependency: libstdc++6
Download-Size: 614 kB
Installed-Size: 2,750 kB
- - - - - -
  Dependency: rhvoice-english
Download-Size: 34.1 MB
Installed-Size: 65.1 MB
- - - - - -
  Dependency: <rhvoice-voice> (virtual package, no size information)
- - - - - -
` = = = = ` = = = = ` = = = = ` = = = = =
Package: rhvoice-english
Download-Size: 34.1 MB
Installed-Size: 65.1 MB
Dependencies:
` = = = = ` = = = = ` = = = = ` = = = = =

Until now I have been successful in implementing only a rudimentary Python created GUI for your rhvoice program (executable). The script is as follows:

` = = = = ` = = = =PYTHON PROGRAM = = = = ` = = = = =
import tkinter as tk
import subprocess

class RHVoiceGUI:
    def __init__(self, root):
        self.root = root
        self.root.title("RHVoice GUI")
        self.root.geometry("400x600")

        # Text Input (Multiline and Scalable)
        self.text_label = tk.Label(root, text="Enter Text:")
        self.text_label.pack(pady=10)
        self.text_input = tk.Text(root, wrap=tk.WORD, width=40, height=10)
        self.text_input.pack(expand=True, fill=tk.BOTH, padx=10, pady=10)

        # Speak Button
        self.speak_button = tk.Button(root, text="Speak", command=self.speak)
        self.speak_button.pack(pady=20)

    def speak(self):
        text = self.text_input.get("1.0", tk.END).strip()
        subprocess.run(['spd-say', '-o', 'rhvoice', text])

if __name__ == "__main__":
    root = tk.Tk()
    app = RHVoiceGUI(root)
    root.mainloop()
` = = = = ` = = = = /PYTHON PROGRAM = = = = ` = = = = =

But I am unable to use the best of rhvoice and improve the output voice quality with a better Python executable as follows:

` = = = = ` = = = = PYTHON PROGRAM= = = = ` = = = = =
import tkinter as tk
import subprocess
import os

class RHVoiceGUI:
    def __init__(self, root):
        self.root = root
        self.root.title("RHVoice GUI")
        self.root.geometry("400x300")

        # Voice Selection
        self.voice_label = tk.Label(root, text="Select Voice:")
        self.voice_label.pack(pady=10)
        self.voice_var = tk.StringVar(value="english-us")
        self.voice_menu = tk.OptionMenu(root, self.voice_var, "english-us", "russian", "spanish")
        self.voice_menu.pack()

        # Text Input (Multiline and Scalable)
        self.text_label = tk.Label(root, text="Enter Text:")
        self.text_label.pack(pady=10)
        self.text_input = tk.Text(root, wrap=tk.WORD, width=40, height=10)
        self.text_input.pack(expand=True, fill=tk.BOTH, padx=10, pady=10)

        # Speak Button
        self.speak_button = tk.Button(root, text="Speak", command=self.speak)
        self.speak_button.pack(pady=20)

    def speak(self):
        text = self.text_input.get("1.0", tk.END).strip()
        voice = self.voice_var.get()
        rhvoice_path = "/usr/bin/RHVoice-client"  # Ensure this is correct

        # Check if RHVoice exists at the given path
        if not os.path.isfile(rhvoice_path):
            print("Error: RHVoice executable not found!")
            return

        # Execute the command
        subprocess.run([rhvoice_path, "-s", voice, text])

if __name__ == "__main__":
    root = tk.Tk()
    app = RHVoiceGUI(root)
    root.mainloop()
` = = = = ` = = = =/PYTHON PROGRAM = = = = ` = = = = =

Where am I going wrong? I am not a degree-holding, professional programmer. This I do as one of my hobbies and interests. I am a physicist.

**Comment by Heavygrass** [comment]
Hi,
You need to use "`def __init__`" instead of "def init".
[https://bobbyhadz.com/blog/python-typeerror-class-takes-no-arguments](https://bobbyhadz.com/blog/python-typeerror-class-takes-no-arguments)
Also indentation is messed up in the post (which is very important for python), better attach python code as files.
Where's "name" in `if name == "main":` coming from?

**Comment by rmcpantoja** [comment]
Hi,
Also, version 1.8 is too old. We are currently in 1.16, which can be obtained in snap or maybe other Linux package managers.

**Reply by Heavygrass** [reply]
> We are currently in 1.16, which can be obtained in snap or maybe other Linux package managers.

@rmcpantoja 
There's 1.8.0 in Ubuntu 24.04.1 repository btw (installed it yesterday).
`apt list --installed |grep rhvoi`

librhvoice-audio2/noble,now 1.8.0+dfsg-3build3 amd64 [installed,automatic]
librhvoice-core7/noble,now 1.8.0+dfsg-3build3 amd64 [installed,automatic]
rhvoice-english/noble,noble,now 1.8.0+dfsg-3build3 all [installed,automatic]
rhvoice/noble,now 1.8.0+dfsg-3build3 amd64 [installed]

How to get 1.16 or smth newer for Ubuntu?
=======================================
And likely a bug in
`RHVoice-client  --version

RHVoice-client  version: none`

**Reply by rmcpantoja** [reply]
> > We are currently in 1.16, which can be obtained in snap or maybe other Linux package managers.
> 
> @rmcpantoja There's 1.8.0 in Ubuntu 24.04.1 repository btw (installed it yesterday). `apt list --installed |grep rhvoi`
> 
> librhvoice-audio2/noble,now 1.8.0+dfsg-3build3 amd64 [installed,automatic] librhvoice-core7/noble,now 1.8.0+dfsg-3build3 amd64 [installed,automatic] rhvoice-english/noble,noble,now 1.8.0+dfsg-3build3 all [installed,automatic] rhvoice/noble,now 1.8.0+dfsg-3build3 amd64 [installed]
> 
> # How to get 1.16 or smth newer for Ubuntu?
> And likely a bug in `RHVoice-client --version
> 
> RHVoice-client version: none`

Hi @Heavygrass, 
Most recent versions are published in [snap](https://snapcraft.io/rhvoice), [source](https://rhvoice.org/linux/)

**Reply by Heavygrass** [reply]
@rmcpantoja Thank you for response.
snap is slow and clogs mounts (lots of issues with snap).
Why not do .deb or anything supported by apt for Ubuntu?

**Comment by zstanecic** [comment]
Or, compile from source. SOurce version is aligned with that in snap. Core functionality is the same.

**Comment by Heavygrass** [comment]
@rajibando Not really my bussiness, but the amount of significance indentation play in Python and lack of static types makes me believe its not the best choice for a long-term project.
I.e. there's simply no compiler to help you with code issues.
I would check C++ or Java.
Lack of static types makes it hard to impossible for anybody else to get involved also.
Python looks like its easy to use, but with the code growing it can make code maintenance very hard.

This project may be a good start:
[https://github.com/infektedpc/Carnival/tree/master](https://github.com/infektedpc/Carnival/tree/master)
